common/workunit/workunit.cpp

```diff
@@ -5113,6 +5167,288 @@ bool CWorkUnitFactory::restoreWorkUnit(const char *base, const char *wuid, bool
+        {
+            //This is used for legacy WU where the /Process/THORNAME/@max is not set. 
+            //From the log names of thor slaves, find out the largest thor slave ID as
+            //the NumberOfThorSlaves. Inside a ZAP reports, the log names of thor slaves 
+            //has the format: 'xxx_thorslave.SLAVE_ID.xxx'. The pStr is pointed at
+            //the end of the '_thorslave.'. The slaveNumStrEnd is pointed at the end
+            //of a SLAVE_ID.
+            unsigned numberOfThorSlaves = 0;
+            VStringBuffer xPath("*%s*", THORSLAVELOGSEARCHSTR);
+            Owned<IFile> zapFileFolder = createIFile(zapWUFolderPath);
+            Owned<IDirectoryIterator> thorSlaveFiles = zapFileFolder->directoryFiles(xPath);
+            ForEach(*thorSlaveFiles)
+            {
+                const char *pStr = strstr(thorSlaveFiles->query().queryFilename(), THORSLAVELOGSEARCHSTR);
+                pStr += strlen(THORSLAVELOGSEARCHSTR);
+                const char *slaveNumStrEnd = strchr(pStr, '.');
+                if (!slaveNumStrEnd)
+                    continue;
+
+                StringBuffer slaveNumStr;
+                slaveNumStr.append(slaveNumStrEnd - pStr, pStr);
+                unsigned slaveNum = atoi(slaveNumStr);
+                if (slaveNum > numberOfThorSlaves)
+                    numberOfThorSlaves = slaveNum;
+            }
+            return numberOfThorSlaves;
+        }
+        void updateWUQueryAssociatedFilesAttrs(const char *localIP)
+        {
+            StringArray filesNotInZAP;
+            IPropertyTree *queryAssociatedTreeRoot = wuTree->queryPropTree("Query/Associated");
+            Owned<IPropertyTreeIterator> itr = queryAssociatedTreeRoot->getElements("File");
+            ForEach (*itr)
+            {
+                IPropertyTree &fileTree = itr->query();
+                const char *fileName = fileTree.queryProp("@filename");
+
+                StringBuffer fileNameWithPath(unzipDir);
+                fileNameWithPath.append(PATHSEPSTR).append(pathTail(fileName));
+                if (checkFileExists(fileNameWithPath)) //Check if the ZAP report contains this file.
+                {
+                    fileTree.setProp("@ip", localIP);
+                    fileTree.setProp("@filename", fileNameWithPath);
+                }
+                else
+                    filesNotInZAP.append(fileName);
+            }
+            ForEachItemIn(i, filesNotInZAP)
+            {
+                VStringBuffer xPath("File[@filename='%s']", filesNotInZAP.item(i));
+                queryAssociatedTreeRoot->removeTree(queryAssociatedTreeRoot->queryPropTree(xPath));
+            }
+        }
+        void getWUAttributes(IWorkUnit *workunit)
+        {
+            wuid.set(workunit->queryWuid());
+            scope.set(workunit->queryWuScope());
+        }
+        IPTree *queryWUPTree()
+        {
+            return wuTree;
+        }
+        IPTree *queryGraphProgressPTree()
```

Comment:
minor: should this be a const method? and should it return const IPTree ?

anywhere that declares a IPTree/IPropertyTree (or other reference/pointer) where it is expected not to change, should be declared so really..

Reply:
It is fine to be a const method. But, returning const IPTree does not work for the LINK(graphProgressTree) inside the import(IPropertyTree *wuTree, IPropertyTree *graphProgressTree).